### PR TITLE
Add formatter for exceptions in plain text

### PR DESCRIPTION
**Pull request by Erikvv**
## Description

Describe your changes in detail.

_Added a formatter for exceptions to an easier-to-read format_
_Removed formatting logic from the zend mvc listener and instead pass the exception to monolog which does a better job. This is a breaking change_

## Motivation and context

Why is this change required? What problem does it solve? _I want to read my logs_

If it fixes an open issue, please link to the issue here (if you write `fixes #num`
or `closes #num`, the issue will be automatically closed when the pull is accepted.)

## How has this been tested?

Please describe in detail how you tested your changes. _Test includes so far as possible. The stacktraces in an Exception/Throwable can't be mocked so this makes an end-to-end test impossible._

Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc.

## Screenshots (if appropriate)

_This is the new format:_

```
---------------------------------------
[Wed, 03 Jan 2018 19:57:35 +0100] RuntimeException: msgn in /var/www/module/Shop/src/Controller/Order.php:25
TotalLubricants\Shop\Controller\Order->orderAction():78
Zend\Mvc\Controller\AbstractActionController->onDispatch(Object(Zend\Mvc\MvcEvent)):322
Zend\EventManager\EventManager->triggerListeners(Object(Zend\Mvc\MvcEvent), Object(Closure)):179
Zend\EventManager\EventManager->triggerEventUntil(Object(Closure), Object(Zend\Mvc\MvcEvent)):106
Zend\Mvc\Controller\AbstractController->dispatch(Object(Zend\Http\PhpEnvironment\Request), Object(Zend\Http\PhpEnvironment\Response)):138
Zend\Mvc\DispatchListener->onDispatch(Object(Zend\Mvc\MvcEvent)):322
Zend\EventManager\EventManager->triggerListeners(Object(Zend\Mvc\MvcEvent), Object(Closure)):179
Zend\EventManager\EventManager->triggerEventUntil(Object(Closure), Object(Zend\Mvc\MvcEvent)):332
Zend\Mvc\Application->run():37
……[Previous Exception] InvalidArgumentException: asdfasdf in /var/www/module/Shop/src/Controller/Order.php:23
……TotalLubricants\Shop\Controller\Order->orderAction():78
……Zend\Mvc\Controller\AbstractActionController->onDispatch(Object(Zend\Mvc\MvcEvent)):322
……Zend\EventManager\EventManager->triggerListeners(Object(Zend\Mvc\MvcEvent), Object(Closure)):179
……Zend\EventManager\EventManager->triggerEventUntil(Object(Closure), Object(Zend\Mvc\MvcEvent)):106
……Zend\Mvc\Controller\AbstractController->dispatch(Object(Zend\Http\PhpEnvironment\Request), Object(Zend\Http\PhpEnvironment\Response)):138
……Zend\Mvc\DispatchListener->onDispatch(Object(Zend\Mvc\MvcEvent)):322
……Zend\EventManager\EventManager->triggerListeners(Object(Zend\Mvc\MvcEvent), Object(Closure)):179
……Zend\EventManager\EventManager->triggerEventUntil(Object(Closure), Object(Zend\Mvc\MvcEvent)):332
……Zend\Mvc\Application->run():37
---------------------------------------
```
_A possible improvement is removing duplicate lines in the previous exception_

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

Go over all the following points, and put an `x` in all the boxes that apply.

Please, please, please, don't send your pull request until all of the boxes are ticked.

- [x] I have read the **[CONTRIBUTING](CONTRIBUTING.md)** document.
- [x] My pull request addresses exactly one patch/feature.
- [x] I have created a branch for this patch/feature.
- [x] Each individual commit in the pull request is meaningful.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] If my change requires a change to the documentation, I have updated it accordingly.
- [x] My code follows the code style of this project.

If you're unsure about any of these, don't hesitate to ask. We're here to help!

### CODE REVIEW

File: src/Formatter/HumanReadableExceptionFormatter.php [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+     * @return string a summary of the argument
+     */
+    protected function formatArgument($argument): string
+    {
+        if (is_int($argument) || is_float($argument)) {
+            return $argument;
+        }
+
+        if (is_string($argument)) {
+            if (strlen($argument) < 80) {
```
Comment (jaapio) [inline src/Formatter/HumanReadableExceptionFormatter.php]: Why would you like to truncate this?

File: src/Formatter/HumanReadableExceptionFormatter.php [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        foreach ($exception->getTrace() as $traceLine) {
+            $lines[] = $this->formatTraceLine($traceLine);
+        }
+
+        if ($exception->getPrevious()) {
+            $previousLines = $this->linesFromException($exception->getPrevious(), true);
+
+            // recursion makes sure deeper nesting has longer prefixes
+            foreach ($previousLines as $previousLine) {
+                $lines[] = '……' . $previousLine;
```
Comment (jaapio) [inline src/Formatter/HumanReadableExceptionFormatter.php]: think I would prefer whitespace indenting here. Makes it more readable?

File: src/Formatter/HumanReadableExceptionFormatter.php [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+    /**
+     * Summarize a variable
+     *
+     * @param mixed $argument anything at all
+     * @return string a summary of the argument
+     */
+    protected function formatArgument($argument): string
+    {
+        if (is_int($argument) || is_float($argument)) {
```
Comment (jaapio) [inline src/Formatter/HumanReadableExceptionFormatter.php]: when you add is_bool here, you can remove the if statements from [R147](https://github.com/polderknowledge/log-module/pull/3/files#diff-8775f8199b48fb7fd9059dc3301ea90aR147)  and do  [R151](https://github.com/polderknowledge/log-module/pull/3/files#diff-8775f8199b48fb7fd9059dc3301ea90aR151) and just do a var_export($argument)

**Review by jaapio — CHANGES_REQUESTED** [review]
I suggested some improvements. Beside that tests are missing.

File: src/Formatter/HumanReadableExceptionFormatter.php (hunk 2 above)
Comment (Erikvv) [reply src/Formatter/HumanReadableExceptionFormatter.php]: Ok I don't really care as long as it's indented

File: src/Formatter/HumanReadableExceptionFormatter.php (hunk 3 above)
Comment (Erikvv) [reply src/Formatter/HumanReadableExceptionFormatter.php]: var_export doesn't work too well with floats
```
php > echo var_export(12.2);
12.199999999999999
php > echo var_export(3.1);
3.1000000000000001
```
In that case I believe the slightly better brevity of var_export doesn't make for better readability

File: src/Formatter/HumanReadableExceptionFormatter.php (hunk 1 above)
Comment (Erikvv) [reply src/Formatter/HumanReadableExceptionFormatter.php]: I don't want e.g. large binaries or 1MB XML strings in my logs.

**Comment by Erikvv** [comment]
Seems it's impossible to write a full test because it's impossible to mock a Throwable because the methods are final. 
 
Nominal typing strikes again

**Comment by Erikvv** [comment]
Added tests in sofar it's possible without overcomplicating things with substrings or regexes.

**Review by jaapio — APPROVED** [review]

**Pull request merged**